File: 687.最长同值路径.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int path = 0;

    int longestUnivaluePath(TreeNode* root) {
        dfs(root);
        return path;
    }

    int dfs(TreeNode* root){
        if(!root)
            return 0;
        int left = dfs(root->left);
        int right = dfs(root->right);
        left = root->left && root->val == root->left->val? left+1: 0;
        right = root->right && root->val == root->right->val? right+1: 0;
        path = max(path, left+right);
        return max(left, right);
    }
};

File: 二叉树/687.最长同值路径.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
 var longestUnivaluePath = function(root) {
    let path = 0;

    const dfs = (root) => {
        if (root == null)
            return 0;
        let left = dfs(root.left);
        let right = dfs(root.right);
        left = root.left != null && root.left.val == root.val? left + 1: 0;
        right = root.right != null && root.right.val == root.val? right + 1: 0;
        path = Math.max(path, left + right);
        return Math.max(left, right);
    }

    dfs(root);

    return path;
};